### PR TITLE
[compiler] Lots of compiler and code size performance improvements

### PR DESCRIPTION
**Pull request by tpoterba**
Necessary for the big aggregate benchmark to pass (albeit more slowly) in benchmarks.

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/ir/Emit.scala [hunk 1]
```diff
@@ -610,8 +610,18 @@ class Emit[C](
       case Void() =>
         Code._empty
 
-      case Begin(xs) =>
-        xs.foreach(x => emitVoid(x))
+      case x@Begin(xs) =>
+        if (!ctx.inLoopCriticalPath.contains(x) && xs.forall(x => !ctx.inLoopCriticalPath.contains(x))) {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/Emit.scala:614]: method splitter doesn't do well with lots of flat children -- splitting manually does better.

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 2]
```diff
@@ -172,17 +173,77 @@ object PruneDeadFields {
           t match {
             case ts: TStruct =>
               val subStructs = children.map(_.asInstanceOf[TStruct])
-              val subFields = ts.fields.map { f =>
-                f -> subStructs.flatMap(s => s.fieldOption(f.name))
+              val fieldArrays = Array.fill(ts.fields.length)(new BoxedArrayBuilder[Type])
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala:176]: This stuff was a major performance hotspot.

File: hail/src/main/scala/is/hail/expr/ir/agg/AggregatorState.scala [hunk 3]
```diff
@@ -36,14 +36,14 @@ trait AggregatorState {
   def deserialize(codec: BufferSpec): (EmitCodeBuilder, Value[InputBuffer]) => Unit
 
   def deserializeFromBytes(cb: EmitCodeBuilder, bytes: SBinaryCode): Unit = {
-    val lazyBuffer = kb.getOrDefineLazyField[MemoryBufferWrapper](Code.newInstance[MemoryBufferWrapper](), (this, "bufferWrapper"))
+    val lazyBuffer = kb.getOrDefineLazyField[MemoryBufferWrapper](Code.newInstance[MemoryBufferWrapper](), ("AggregatorStateBufferWrapper"))
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/AggregatorState.scala:39]: we weren't caching the way we thought.

File: hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala [hunk 4]
```diff
@@ -339,47 +339,56 @@ object Extract {
...
     val (initOps, pAggSigs) = ab.result().unzip
     val rt = TTuple(initOps.map(_.aggSig.resultType): _*)
     ref._typ = rt
 
     Aggs(postAgg, Begin(initOps), addLets(Begin(seq.result()), let.result()), pAggSigs)
   }
 
-  private def extract(ir: IR, ab: BoxedArrayBuilder[(InitOp, PhysicalAggSig)], seqBuilder: BoxedArrayBuilder[IR], letBuilder: BoxedArrayBuilder[AggLet], result: IR, r: RequirednessAnalysis, isScan: Boolean): IR = {
-    def extract(node: IR): IR = this.extract(node, ab, seqBuilder, letBuilder, result, r, isScan)
+  private def extract(ir: IR, ab: BoxedArrayBuilder[(InitOp, PhysicalAggSig)], seqBuilder: BoxedArrayBuilder[IR], letBuilder: BoxedArrayBuilder[AggLet], memo: mutable.Map[IR, Int], result: IR, r: RequirednessAnalysis, isScan: Boolean): IR = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala:351]: Here we cache agg IRs by identity, which deduplicates counts.

File: hail/src/main/scala/is/hail/lir/Blocks.scala [hunk 5]
```diff
@@ -3,13 +3,17 @@ package is.hail.lir
 class Blocks(
   blocks: Array[Block]
 ) extends IndexedSeq[Block] {
-  private val blockIdx = blocks.iterator.zipWithIndex.toMap
+  private val blockIdx = {
```
Comment (tpoterba) [inline hail/src/main/scala/is/hail/lir/Blocks.scala:6]: java map is faster than scala.

File: hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala [hunk 6]
```diff
@@ -93,7 +93,7 @@ sealed abstract class BaseTypeWithRequiredness {
       throw new AssertionError(
         s"children lengths differed ${children.length} ${newChildren.length}. ${children} ${newChildren} ${this}")
     }
-    children.zip(newChildren).foreach { case (r1, r2) =>
+    (children, newChildren).zipped.foreach { case (r1, r2) =>
```
[the remark this answers is not in the document]
Comment (patrick-schultz) [reply hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala:96]: Oh cool, I hadn't realized how `zipped` was implemented. The argument to `foreach` is actually a 2-ary function, so you don't even need the `case`. That may be a performance improvement, but also makes it immediately clear that we're not allocating tuples. (Likewise for the other uses of `zipped` in this file.)

File: hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala [hunk 7]
```diff
@@ -361,7 +362,7 @@ sealed abstract class RBaseStruct extends TypeWithRequiredness {
   def size: Int = fields.length
   val children: Seq[TypeWithRequiredness] = fields.map(_.typ)
   def _unionLiteral(a: Annotation): Unit =
-    children.zip(a.asInstanceOf[Row].toSeq).foreach { case (r, f) => r.unionLiteral(f) }
+    children.iterator.zip(rowIterator(a.asInstanceOf[Row])).foreach { case (r, f) => r.unionLiteral(f) }
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala:365]: `zipped`?

**Review by patrick-schultz — DISMISSED** [review]

File: hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala (hunk 7 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/types/TypeWithRequiredness.scala:365]: oops, missed one, thanks.

File: hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala [hunk 8]
```diff
@@ -979,15 +1040,15 @@ object PruneDeadFields {
         val bodyEnv = memoizeValueIR(body,
           requestedType.asInstanceOf[TStream].elementType,
           memo)
-        val valueTypes = names.zip(as).map { case (name, a) =>
+        val valueTypes = (names, as).zipped.map { case (name, a) =>
```
Comment (patrick-schultz) [inline hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala:1043]: As discussed in person, function literals with `case` still force a tuple to be allocated.

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**